### PR TITLE
Fix child device

### PR DESCRIPTION
**Pull request by SkafteNicki**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## What does this PR do?

Fixes https://github.com/PyTorchLightning/metrics/issues/531
Currently if a metric is a child of a module, the `device` property (only the property, the metric states are still correctly updated) of the metric will not get properly update (only if `.cuda()`, `.cpu()`, `to()` is directly called on the metric) when calling `.cuda()` on the module. This PR fixes it.

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by cowwoc** [comment]
@SkafteNicki Thank you for working on this fix!

**Comment by SkafteNicki** [comment]
@cowwoc if you have the time, please confirm that this fixes your original issue :]

File: torchmetrics/metric.py [hunk 1]
```diff
@@ -515,6 +515,10 @@ def _apply(self, fn: Callable) -> Module:
                     "Expected metric state to be either a Tensor" f"or a list of Tensor, but encountered {current_val}"
                 )
 
+        # make sure to update the device attribute
+        # if the dummy tensor moves device by fn function we should also update the attribute
+        self._device = fn(torch.zeros(1, device=self.device)).device
```
Comment (justusschock) [inline torchmetrics/metric.py:520]: Why is this needed at all. Shouldn't this be handled by our wrapped `.to` function?

Comment (SkafteNicki) [reply torchmetrics/metric.py:520]: Because if you call `.to` on a parent module, it does actually not call the `.to` method of the child modules but it instead call `_apply` method. This change actually makes the wrapped `cuda`, `cpu`, `to` we have right now obsolete.

Comment (justusschock) [reply torchmetrics/metric.py:520]: I see, thanks :) But then do we also want to get the dtype there or is that something we only want with an explicit call of `.to` etc?

Comment (SkafteNicki) [reply torchmetrics/metric.py:520]: the `dtype` casting was actually remove in a earlier PR, because users were worried that if they run mixed-precision training that their metrics would be automatically casted to half precision given them wrong results. It can still be done but you need to explicitly call the special method `set_type`.

Comment (justusschock) [reply torchmetrics/metric.py:520]: Should we then extend this PR to remove the legacy wrappings for the device changes? Since this won't change behaviour or API it should be fine, right?

**Review by justusschock — APPROVED** [review]

**Comment by an1lam** [comment]
Thanks @SkafteNicki!

**Review by Borda — APPROVED** [review]

File: torchmetrics/metric.py (hunk 1 above)
Comment (SkafteNicki) [reply torchmetrics/metric.py:478]: @justusschock you are right, it is removed now :]
